rfcs/text/0000-create-file-mach-o.md

```diff
@@ -0,0 +1,155 @@
+|   file.mach-o.headers.flags                |   keyword  |   Flags set in the Mach-O header.                                           |
+|   file.mach-o.segments                     |   object   |   Segment information for the file.                                         |
+|   file.mach-o.segments.vmaddr              |   keyword  |   Memory address of this segment.                                           |
+|   file.mach-o.segments.name                |   keyword  |   Name of this segment.                                                     |
+|   file.mach-o.segments.vmsize              |   keyword  |   Memory size of this segment.                                              |
+|   file.mach-o.segments.fileoff             |   keyword  |   File offset of this segment.                                              |
+|   file.mach-o.segments.filesize            |   keyword  |   Amount of memory to map from the file.                                    |
+|   file.mach-o.segments.sections            |   object   |   Section information for the segment of the file.                          |
+|   file.mach-o.segments.sections.flags      |   keyword  |   Section flags for the segment of the file.                                |
+|   file.mach-o.segments.sections.name       |   keyword  |   Section name for the segment of the file.                                 |
+|   file.mach-o.segments.sections.type       |   keyword  |   Section type for the segment of the file.                                 |
+|   file.mach-o.signature                    |   object   |   Signature information for the file.                                       |
+|   file.mach-o.signature.candidate_cd_hash  |   keyword  |   Code Digest (CD) SHA256 hash of the first 20-bytes of the file.           |
+|   file.mach-o.signature.team_identifier    |   keyword  |   Team identifier of the code signing certificate.                          |
+|   file.mach-o.signature.sealed_resources   |   long     |   Version of the resource envelope for the code signing certificate.        |
+|   file.mach-o.signature.cms_digest         |   keyword  |   Cryptographic Message Syntax (CMS) hash of the code signing certificate.  |
+|   file.mach-o.signature.cms_digest_type    |   keyword  |   Cryptographic Message Syntax (CMS) type of the code signing certificate.  |
+|   file.mach-o.signature.status             |   keyword  |   Verification information for the code signing certificate.                |
+|   file.mach-o.signature.fingerprint        |   keyword  |   MD5 digest of the der-encoded certificate information.                    |
+|   file.mach-o.executable                   |   object   |   Information about the executable segment for the file.                    |
+|   file.mach-o.executable.segment_base      |   keyword  |   Executable segment base size.                                             |
```

Comment:
wondering why specifically calling out the executable segment here rather than keeping this information for each segment and dropping `file.mach-o.executable`?

Reply:
Apologies, I don't have a great answer here. In looking at the data of a single file, here's what I'm seeing for `executable`:
```
...
"Executable Segment base": "0",
"Executable Segment limit": "49152",
"Executable Segment flags": "0x0",
...
```
Compared to what I'm seeing for the `segments` section:
```
 "segments": [
        {
          "vmaddr": "0x0",
          "name": "__TEXT",
          "vmsize": "0xc000",
          "fileoff": "0x0",
          "filesize": "0xc000",
          "sections": [
            {
              "flags": [
                "SECTION_ATTRIBUTES_USR",
                "SECTION_ATTRIBUTES_SYS"
              ],
              "name": "__text",
              "type": "S_REGULAR"
```

Of note, in the second example (from the same file), `file.mach-o.segments.sections.flags:SECTION_ATTRIBUTES_USR` and `file.mach-o.executable.segment_flags:0x0`...so it appears to represent different parts of the file.

Would you prefer to see something like this?
```
file.mach-o.segments.executable.base
file.mach-o.segments.executable.limit
file.mach-o.segments.executable.flags
```

Reply:
So, from the looks of:

```
"Executable Segment base": "0",
"Executable Segment limit": "49152",
"Executable Segment flags": "0x0",
```

it seems that what's being described is the offset, size, and flags of the segment with executable code in it, namely the `__TEXT` segment (which will have a number of sections in it). Here's a [reference](https://github.com/aidansteele/osx-abi-macho-file-format-reference#table-1-the-sections-of-a__textsegment).

I would think that it would probably make more sense just to model this as something like:

```
file.mach-o.segments.name = "__TEXT"
file.mach-o.segments.offset = 0
file.mach-o.segments.size = 49152
file.mach-o.segments.flags = []
```

Which would be pretty much identical to with something like a `__DATA` or `__IMPORT` segment.

Also, note that for:

> Of note, in the second example (from the same file), file.mach-o.segments.sections.flags:SECTION_ATTRIBUTES_USR and file.mach-o.executable.segment_flags:0x0...so it appears to represent different parts of the file.

the former is referring to the section flags of the `__text` section inside of the `__TEXT` segment (wonderful names I know), while the latter is referring to the flags of the `__TEXT` segment itself.

Reply:
Thanks for the explanation. I like this approach - seems cleaner
```
file.mach-o.segments.name = "__TEXT"
file.mach-o.segments.offset = 0
file.mach-o.segments.size = 49152
file.mach-o.segments.flags = []
```
I'll do this.